### PR TITLE
Support kubernetes version 22+

### PR DESCRIPTION
**Pull request by mandopaloooza**
* Fix issue #141 
* Change CRD version to `apiextensions.k8s.io/v1`
* Generate JSON schema using `apischema` and include in the CRD spec in order to comply with requirements for structural schema

```
root@9a0761bd7c18:/test# /root/run.sh dump-crd
2022-05-24 19:26:25,723 [INFO] [logger] Set log level to info
2022-05-24 19:26:25,808 [INFO] Generating entity Policy for path /policies
2022-05-24 19:26:25,900 [INFO] Generating entity Condition for path /conditions
2022-05-24 19:26:25,944 [INFO] Generating entity Entitlement for path /entitlements
2022-05-24 19:26:25,988 [INFO] Generating entity RingfenceRule for path /ringfence-rules
2022-05-24 19:26:26,017 [INFO] Generating entity Appliance for path /appliances
2022-05-24 19:26:26,244 [WARNING] Attrib peerInterface is deprecated but required by openapi spec
2022-05-24 19:26:26,283 [INFO] Generating entity Site for path /sites
2022-05-24 19:26:26,367 [INFO] Generating entity IpPool for path /ip-pools
2022-05-24 19:26:26,403 [INFO] Generating entity IdentityProvider for path /identity-providers
2022-05-24 19:26:26,466 [INFO] Generating entity LocalUser for path /local-users
2022-05-24 19:26:26,491 [INFO] Generating entity AdministrativeRole for path /administrative-roles
2022-05-24 19:26:26,527 [INFO] Generating entity MfaProvider for path /mfa-providers
2022-05-24 19:26:26,559 [INFO] Generating entity AdminMfaSettings for path /admin-mfa-settings
2022-05-24 19:26:26,571 [INFO] Generating entity TrustedCertificate for path /trusted-certificates
2022-05-24 19:26:26,652 [INFO] Generating entity CriteriaScripts for path /criteria-scripts
2022-05-24 19:26:26,673 [INFO] Generating entity DeviceScript for path /device-scripts
2022-05-24 19:26:26,702 [INFO] Generating entity EntitlementScript for path /entitlement-scripts
2022-05-24 19:26:26,735 [INFO] Generating entity ApplianceCustomization for path /appliance-customizations
2022-05-24 19:26:26,759 [INFO] Generating entity GlobalSettings for path /global-settings
2022-05-24 19:26:26,780 [INFO] Generating entity ClientConnection for path /client-connections
2022-05-24 19:26:27,286 [INFO] [dump-crd] File 2022-05-24_19-26-crd.yml generated with CRD definitions
```

```
$ kubectl apply -f 2022-05-24_19-26-crd.yml
customresourcedefinition.apiextensions.k8s.io/policies.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/conditions.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/entitlements.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/ringfencerules.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/appliances.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/sites.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/ippools.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/identityproviders.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/localusers.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/administrativeroles.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/mfaproviders.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/adminmfasettingss.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/trustedcertificates.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/criteriascriptss.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/devicescripts.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/entitlementscripts.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/appliancecustomizations.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/globalsettingss.beta.appgate.com created
customresourcedefinition.apiextensions.k8s.io/clientconnections.beta.appgate.com created
```

### CODE REVIEW

**Comment by mandopaloooza** [comment]
Build needs fix from https://github.com/appgate/sdp-operator/pull/148

File: appgate/openapi/openapi.py [hunk 1]
```diff
@@ -118,7 +124,61 @@ def entity_names(
 
 
 def generate_crd(entity: Type, short_names: Dict[str, str]) -> str:
+    prev_default_object_fields = settings.default_object_fields
+
+    def attrs_fields(cls: type) -> Optional[Sequence[ObjectField]]:
+        if hasattr(cls, "__attrs_attrs__"):
+            obj: List[ObjectField] = []
+            for a in getattr(cls, "__attrs_attrs__"):
```
Comment (thomascellerier) [inline appgate/openapi/openapi.py:132]: You can use the `attrs.fields` function for this, see https://www.attrs.org/en/stable/api.html#attrs.fields

File: appgate/openapi/openapi.py [hunk 2]
```diff
@@ -118,7 +124,61 @@ def entity_names(
 
 
 def generate_crd(entity: Type, short_names: Dict[str, str]) -> str:
+    prev_default_object_fields = settings.default_object_fields
+
+    def attrs_fields(cls: type) -> Optional[Sequence[ObjectField]]:
+        if hasattr(cls, "__attrs_attrs__"):
```
Comment (thomascellerier) [inline appgate/openapi/openapi.py:130]: You can use `attrs.has` to check if it is an attrs class https://www.attrs.org/en/stable/api.html#attrs.has

File: tests/test_crd.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+from appgate.openapi.openapi import generate_crd, generate_api_spec, SPEC_DIR
+
+
+def test_generate_crd():
+    for version in ["v12", "v13", "v14", "v15", "v16"]:
+        open_api = generate_api_spec(Path(SPEC_DIR).parent / version)
+        for k, v in open_api.entities.items():
+            crd = generate_crd(open_api.entities[k].cls, {})
+            assert crd
```
Comment (thomascellerier) [inline tests/test_crd.py:11]: Could we check the actual generated schema here? At least for a few entities.

File: appgate/openapi/openapi.py [hunk 4]
```diff
@@ -128,23 +188,31 @@ def generate_crd(entity: Type, short_names: Dict[str, str]) -> str:
...
-    """
-    TODO: Iterate over all attributes here to generate the spec
-    for a in entity.__attrs_attrs__:
-        spec = a.metadata['spec']
-        str[a.name] = spec()
-    """
+
+    # Register yaml representers for apischema custom types AliasedStr and JsonType
+    def str_representer(dumper, data):
+        return dumper.represent_scalar(u'tag:yaml.org,2002:str', data)
```
Comment (thomascellerier) [inline appgate/openapi/openapi.py:212]: No need to use the u prefix, thats only useful for python2 compat.

**Review by thomascellerier — COMMENTED** [review]
Looks good in general but a few small comments and some questions.

How does this affect the operator beyond CRD generation?
Does the operator then require CRDs using `apiextensions.k8s.io/v1` or will it still work against `apiextensions.k8s.io/v1beta`? What does it mean in terms of upgrading?

File: appgate/openapi/openapi.py [hunk 5]
```diff
@@ -118,7 +124,61 @@ def entity_names(
...
+    settings.default_object_fields = attrs_fields
+
     name, singular_name, plural_name, short_name = entity_names(entity, short_names)
+    schema = deserialization_schema(entity)
+
+    # apischema deserializes nullable properties into [JsonType.*, JsonType.NULL]
+    def replace_nullable_type(obj: dict) -> dict:
+        for k, v in obj.items():
+            if isinstance(v, dict):
+                obj[k] = replace_nullable_type(v)
```
Comment (lepiolet) [inline appgate/openapi/openapi.py:149]: remember ... python is crap when doing recursion :D

File: tests/test_crd.py (hunk 3 above)
Comment (lepiolet) [reply tests/test_crd.py:11]: also make sure that we respect the k8s vs appgate serialization in the schema, some fields are not supposed to be present in k8s so they dont be in the schema.

**Comment by lepiolet** [comment]
> How does this affect the operator beyond CRD generation?

It should not, that API version is only used when generating the CRD not the instances.
For the instances we need to implement the versioning we have discussed.

> Does the operator then require CRDs using apiextensions.k8s.io/v1 or will it still work against apiextensions.k8s.io/v1beta? What does it mean in terms of upgrading?

From https://kubernetes.io/docs/reference/using-api/deprecation-guide/ 
```
The apiextensions.k8s.io/v1beta1 API version of CustomResourceDefinition is no longer served as of v1.22.```

so v1beta1 won't work when >= v1.22

Upgrading should be as simple as adding the new CRDs since:

 ```All existing persisted objects are accessible via the new API```

File: tests/test_crd.py (hunk 3 above)
Comment (mandopaloooza) [reply tests/test_crd.py]: @AitorATuin Do you remember which entity in particular I should watch out for extra fields?

Comment (lepiolet) [reply tests/test_crd.py]: https://github.com/appgate/sdp-operator/blob/main/appgate/attrs.py#L170

See that @mandopaloooza we use the attr read_only/write_only to know if it needs to be used or not, so in your case ignore the fields read_only

File: tests/resources/crd/v12/appliance.yaml [hunk 6]
```diff
@@ -981,13 +969,6 @@ spec:
                 items:
                   type: string
                 type: array
-              updated:
```
Comment (lepiolet) [inline tests/resources/crd/v12/appliance.yaml]: yeah, now it looks better. All this stuff is pure controller info.

**Review by lepiolet — APPROVED** [review]
LGTM

**Pull request merged**